sycl/source/detail/scheduler/commands.cpp

```diff
@@ -169,12 +169,26 @@ static std::string commandToName(Command::CommandType Type) {
 }
 #endif
 
-static std::vector<RT::PiEvent>
-getPiEvents(const std::vector<EventImplPtr> &EventImpls) {
+std::vector<RT::PiEvent>
+Command::getPiEvents(const std::vector<EventImplPtr> &EventImpls) const {
   std::vector<RT::PiEvent> RetPiEvents;
   for (auto &EventImpl : EventImpls) {
-    if (EventImpl->getHandleRef() != nullptr)
-      RetPiEvents.push_back(EventImpl->getHandleRef());
+    if (EventImpl->getHandleRef() == nullptr)
+      continue;
+
+    // Do not add redundant event dependencies for in-order queues.
+    // At this stage dependency is definitely pi task and need to check if
+    // current one is a host task. In this case we should not skip pi event due
+    // to different sync mechanisms for different task types on in-order queue.
+    auto WorkerQueue = getWorkerQueue();
+    if (EventImpl->getWorkerQueue() == WorkerQueue &&
+        WorkerQueue->has_property<property::queue::in_order>() &&
+        (MType != CommandType::RUN_CG /* host task has this type also */ ||
+         (static_cast<const ExecCGCommand *>(this))->getCG().getType() !=
+             CG::CGTYPE::CodeplayHostTask))
```

Comment:
NIT: it seems like this one deserves a macro or static function as we have several places for such checks.

Reply:
will fix

Reply:
fixed in commit https://github.com/intel/llvm/pull/5901/commits/a0c087b43012eb41111cefb0bf79b9e659bce928